### ISSUE TITLE
Type mismatch in dynamodb LastEvaluatedKey

### ISSUE THREAD

**Issue by dacevedo12** [youtype/mypy_boto3_builder#339]
**Describe the bug**
While paginating in dynamo, it is common to send LastEvaluatedKey from the previous response, but the type hints between that previous response and the input of the next one are incompatible.

```python
response = await table_resource.query(...)
while response.get("LastEvaluatedKey"):
            response = await table_resource.query(
                ...,
                ExclusiveStartKey=response.get("LastEvaluatedKey"),
            )
```

`Argument "ExclusiveStartKey" to "query" of "Table" has incompatible type "dict[str, bytes | bytearray | str | int | Decimal | <8 more items> | None] | None"; expected "Mapping[str, bytes | bytearray | str | int | Decimal | <8 more items> | None]" `

TL;DR QueryOutputTableTypeDef declares it as `NotRequired[dict[str, TableAttributeValueTypeDef]]` but QueryInputTableQueryTypeDef expects `NotRequired[Mapping[str, TableAttributeValueTypeDef]]`

for some reason dict is not equivalent to mapping

**Additional context**
13.3.0.post1

python 3.11

**Comment by vemel** [comment]
Hello! `dict` argument is accepted as `Mapping`. The issue is that `response.get("LastEvaluatedKey")` can potentially be `None`, this is what type checker complains about. You are checking that it is not `None` above, but type checker does not accept this kind of type narrowing. Use a variable instead:

```python
response = await table_resource.query(...)

last_evaluated_key = response.get("LastEvaluatedKey")
# type here is dict[str, TableAttributeValueTypeDef] | None

while last_evaluated_key:
     # type here is dict[str, TableAttributeValueTypeDef], because of the condition above
    response = await table_resource.query(
        ...,
        ExclusiveStartKey=last_evaluated_key,
    )
    last_evaluated_key = response.get("LastEvaluatedKey")
```

**Comment by dacevedo12** [comment]
You're right! Thanks a lot
